apps/passport-client/src/zapp/useZappServer.ts

```diff
@@ -0,0 +1,185 @@
+          console.log("invoking function", functionToInvoke, message.args);
+          try {
+            const result = await functionToInvoke.apply(object, message.args);
+            console.log(result);
+            port.postMessage({
+              type: RPCMessageType.ZUPASS_CLIENT_INVOKE_RESULT,
+              result,
+              serial: message.serial
+            } satisfies RPCMessage);
+          } catch (e) {
+            port.postMessage({
+              type: RPCMessageType.ZUPASS_CLIENT_INVOKE_ERROR,
+              serial: message.serial,
+              error: getErrorMessage(e)
+            } satisfies RPCMessage);
+          }
+        } else {
+          throw new Error("Function not found");
+        }
+      } catch (error) {
+        port.postMessage({
+          type: RPCMessageType.ZUPASS_CLIENT_INVOKE_ERROR,
+          error: getErrorMessage(error)
+        });
+      }
+    }
+  });
+}
+
+// In real usage, this would be the user's Semaphore v4 key
+const MAGIC_PRIVATE_KEY =
+  "00112233445566778899AABBCCDDEEFF00112233445566778899aabbccddeeff";
+
+export function useZappServer(): void {
+  const context = useStateContext();
+
+  useEffect(() => {
+    window.addEventListener("message", async (event) => {
+      let port: MessagePort;
+      // @todo: handle repeat calls?
+      console.log(event);
+      const data = WindowMessageSchema.safeParse(event.data);
+      if (!data.success) {
+        return;
+      }
+      const msg = data.data;
+      console.log("passport-client received message ", msg);
+      if (msg.type === WindowMessageType.ZUPASS_CLIENT_CONNECT) {
+        const origin = event.origin;
+        port = event.ports[0];
+        port.start();
+
+        if (!context.getState().self) {
+          console.log("user not logged in");
+          // User is not logged in
+          postRPCMessage(port, {
+            type: RPCMessageType.ZUPASS_CLIENT_SHOW
+          });
+
+          await new Promise<void>((resolve) => {
+            const unlisten = context.stateEmitter.listen((state) => {
+              if (state.self) {
+                postRPCMessage(port, {
+                  type: RPCMessageType.ZUPASS_CLIENT_HIDE
+                });
+                unlisten();
+                resolve();
+              }
+            });
+          });
+        }
+
+        const zapps = context.getState().pcds.getAllPCDsInFolder("Zapps");
+        let zapp = zapps.filter(isPODPCD).find((zapp) => {
+          return Object.entries(zapp.claim.entries).find(([key, entry]) => {
+            if (
+              key === "origin" &&
+              entry.type === "string" &&
+              entry.value === origin
+            ) {
+              return true;
+            }
+            return false;
+          });
+        });
+
+        let approved = !!zapp;
+
+        if (!zapp) {
+          approved = confirm(`Allow ${msg.zapp.name} to connect to Zupass?`);
```

Comment:
👍

Reply:
for the V0 we should add "This is highly experimental -- please make sure that you trust this website."